libs/zglfw/src/zglfw.zig

```diff
@@ -233,6 +244,147 @@ pub const CursorMode = enum(i32) {
+
+    pub const maximum_supported = 16;
+
+    pub fn init(jid: u4) ?Joystick {
+        const self: Joystick = .{ .jid = jid };
+        if (!self.present()) {
+            return null;
+        }
+        return self;
+    }
+
+    pub fn present(self: Joystick) bool {
+        return glfwJoystickPresent(@intCast(i32, self.jid)) == @boolToInt(true);
+    }
+    extern fn glfwJoystickPresent(jid: i32) i32;
+
+    pub fn getGUID(self: Joystick) [*:0]const u8 {
+        return glfwGetJoystickGUID(@intCast(i32, self.jid));
+    }
+    extern fn glfwGetJoystickGUID(jid: i32) [*:0]const u8;
+
+    const empty_axes: [0]f32 = undefined;
+    pub fn getAxes(self: Joystick) []const f32 {
+        var count: i32 = undefined;
+        const state = glfwGetJoystickAxes(@intCast(i32, self.jid), &count);
+        if (count == 0) {
+            return &empty_axes;
+        }
+        return state[0..@intCast(usize, count)];
+    }
+    extern fn glfwGetJoystickAxes(jid: i32, count: *i32) [*]const f32;
+
+    const empty_buttons: [0]ButtonAction = undefined;
+    pub fn getButtons(self: Joystick) []const ButtonAction {
+        var count: i32 = undefined;
+        const state = glfwGetJoystickButtons(@intCast(i32, self.jid), &count);
+        if (count == 0) {
+            return &empty_buttons;
+        }
+        return @ptrCast([]const ButtonAction, state[0..@intCast(usize, count)]);
+    }
+    extern fn glfwGetJoystickButtons(jid: i32, count: *i32) [*]const u8;
+
+    fn isGamepad(self: Joystick) bool {
+        return glfwJoystickIsGamepad(@intCast(i32, self.jid)) == @boolToInt(true);
+    }
+    extern fn glfwJoystickIsGamepad(jid: i32) i32;
+
+    pub fn gamepad(self: Joystick) ?Gamepad {
+        return if (self.isGamepad()) .{ .jid = self.jid } else null;
+    }
+};
+//--------------------------------------------------------------------------------------------------
+//
+// Gamepad
+//
+//--------------------------------------------------------------------------------------------------
+pub const Gamepad = struct {
+    pub const Axis = enum(u3) {
+        left_x = 0,
+        left_y = 1,
+        right_x = 2,
+        right_y = 3,
+        left_trigger = 4,
+        right_trigger = 5,
+
+        const last = Axis.right_trigger;
+    };
+
+    pub const Button = enum(u4) {
+        a = 0,
+        b = 1,
+        x = 2,
+        y = 3,
+        left_bumper = 4,
+        right_bumper = 5,
+        back = 6,
+        start = 7,
+        guide = 8,
+        left_thumb = 9,
+        right_thumb = 10,
+        dpad_up = 11,
+        dpad_right = 12,
+        dpad_down = 13,
+        dpad_left = 14,
+
+        const last = Button.dpad_left;
+
+        const cross = Button.a;
+        const circle = Button.b;
+        const square = Button.x;
+        const triangle = Button.y;
+    };
+
+    pub const State = extern struct {
+        buttons: [15]Joystick.ButtonAction,
+        axes: [6]f32,
+    };
+
+    jid: u4,
+
+    pub fn init(jid: u4) ?Gamepad {
```

Comment:
`init()` suggests that the object is constructed and that `deinit()` should be called to destruct it. Let's remove this function.

```zig
if (zglfw.getJoystick(1)) |joystick| {
    if (joystick.asGamepad()) |gamepad| {
    }
}
```